use mstrlines to cleanup MS-Transferor rules for announced wfs

hassan11196 opened the pull request:
Fixes #12328

#### Status
not-tested

#### Description
Use the  `mstrlines` cleanup pipeline to cleanup MS-Transferor rules for announced wfs.
UPDATE: execute the input data cleanup as soon as the workflow is in `announced` status and all the Disk and Tape rules have been created by MSOutput. Note that the other pipelines (output data) will only be executed once the workflow is ready to be archived.

#### Is it backward compatible (if not, which system it affects?)
YES

#### Related PRs
N/A

#### External dependencies / deployment changes
No

hassan11196 committed:
use mstrlines to cleanup MSTransferor input rules for announced wfs

hassan11196 commented:
Hi @todor-ivanov, Please point me to file you mentioned that should have the change.
Thanks a lot.

todor-ivanov commented:
hi @hassan11196 , the place of change is correct indeed. I was too quick in reading it during the meeting. The rest of my comments I'll leave in the review.

todor-ivanov commented on src/python/WMCore/MicroService/MSRuleCleaner/MSRuleCleaner.py:
```diff
@@ -317,9 +317,21 @@ def _dispatchWflow(self, wflow):
+                    msg = "%s: General error from pipeline while cleaning Transfer rukes. Workflow: %s. Error:  \n%s. " % (pline.name, wflow['RequestName'], str(ex))
+                    msg += "\nWill retry again in the next cycle."
+                    self.logger.exception(msg)
+                    self._checkStatusAdvanceExpired(wflow, additionalInfo=msg)
```
The call to `self._checkStatusAdvanceExpired` is redundant here. It already happens on line 322. You may safely remove this one here.

todor-ivanov commented on src/python/WMCore/MicroService/MSRuleCleaner/MSRuleCleaner.py:
```diff
@@ -317,9 +317,21 @@ def _dispatchWflow(self, wflow):
+                    self.logger.exception(msg)
+                    self._checkStatusAdvanceExpired(wflow, additionalInfo=msg)
+                    continue
+                if wflow['CleanupStatus'][pline.name]:
```
If you update the cleanup counters, while only calling a subset of the cleanup lines, on the next stage, while traversing the archival pipeline, the workflow will be considered cleaned and will trigger a premature archival, before having it passed through all the cleanup lines. This is a consequence from the fact that  the `archival` step and the `cleanup` steps live in the same micro service and are called sequentially, but that's another story and we can live with what we have already. 

Anyway, since you are calling a partial and in advance cleanup of only input rules, the way to avoid this is by NOT updating the cleanup counters. So please just remove these two lines here.

todor-ivanov reviewed (changes requested):
Hi @hassan11196 thanks for the PR, please find my comments inline. 

Your change is basically correct and upon addressing my remarks, the micro service will indeed behave the way how you expect it to, but if that is to be considered the correct behavior, will be decided upon the conversation that we are supposed to have during our internal meeting today. We will let you know the outcome.

amaltaro commented on src/python/WMCore/MicroService/MSRuleCleaner/MSRuleCleaner.py:
```diff
@@ -317,9 +317,18 @@ def _dispatchWflow(self, wflow):
             #       We still need some proper logging for them.
             msg = "Skipping workflow: %s - tape transfers are not yet completed." % wflow['RequestName']
-            msg += " Will retry again in the next cycle."
+            msg += " Will retry again in the next cycle. deleting ONLY the input MS-Transferor rules as wf is announced."
```
How about the following log record here:
```
msg += "Workflow in 'announced' state, hence proceeding only with MSTransferor / input data removal. "
msg += " Will retry the remaining in the next cycle.
```

amaltaro commented on src/python/WMCore/MicroService/MSRuleCleaner/MSRuleCleaner.py:
```diff
@@ -317,9 +317,18 @@ def _dispatchWflow(self, wflow):
+                try: 
+                    pline.run(wflow)
+                except Exception as ex:
+                    msg = "%s: General error from pipeline while cleaning input MS-Transferor rules. Workflow: %s. Error:  \n%s. " % (pline.name, wflow['RequestName'], str(ex))
```
Please break this into shorter (~<100 chars) lines. And use MSTransferor instead of MS-Transferor, just for the sake of consistency.

amaltaro reviewed (changes requested):
Thanks Ahmed. I left a few comments along the code.

However, I have two more important remarks to make (one made in the issue itself)
1) we need to define what the conditions we want to have for this input data removal (I suggest to define it in the issue, then update this PR)
2) ideally, we should have a flag to say that input data has been already removed. Otherwise - with the current Tape issues - we would be retrying these MSTransferor pipelines **every** cycle for thousands of workflows. In principle it does not break anything, but it would be wasting tons of resources (CPU, IO).

hassan11196 commented:
Hi @amaltaro, Thanks for the review 
regarding 
> 1. we need to define what the conditions we want to have for this input data removal (I suggest to define it in the issue, then update this PR)
I have replied here, https://github.com/dmwm/WMCore/issues/12328#issuecomment-2804964902. 
our goal is to clean out the input rules as soon as the wf is announced. I added the condition to the condition checking announced status and TransferTape=false, i.e [link](https://github.com/hassan11196/WMCore/blob/fd41ec75f981cc0ba6123fe01ed64345bbd80f8d/src/python/WMCore/MicroService/MSRuleCleaner/MSRuleCleaner.py#L313)

TransferDone=true as I understand means the output data placement rules are created by MS-Output. I think in normal conditions yes this is ideal, we should keep the input rules until at least the output data rules are created.

If you forsee issues with this I can move it to the upper check that is [link](https://github.com/hassan11196/WMCore/blob/fd41ec75f981cc0ba6123fe01ed64345bbd80f8d/src/python/WMCore/MicroService/MSRuleCleaner/MSRuleCleaner.py#L306)

> 2. ideally, we should have a flag to say that input data has been already removed. Otherwise - with the current Tape issues - we would be retrying these MSTransferor pipelines **every** cycle for thousands of workflows. In principle it does not break anything, but it would be wasting tons of resources (CPU, IO).


I thought so yes, because right now this will keep on trying to delete input rules for all announced wf in every cycle, but where do I store this information that for the wfs the input rules are cleaned? MSRuleCleanor seems to be processing all wfs by statuses defined here, [link](https://github.com/hassan11196/WMCore/blob/e278780d004f56ebe4234b946921a207ae048ec4/src/python/WMCore/MicroService/MSCore/MSManager.py#L160). 
What do you suggest?

amaltaro commented on src/python/WMCore/MicroService/MSRuleCleaner/MSRuleCleaner.py:
```diff
@@ -317,9 +317,21 @@ def _dispatchWflow(self, wflow):
+                    pline.run(wflow)
+                except Exception as ex:
+                    msg = f"{pline.name}: General error from pipeline"
+                    msg += "when cleaning input MSTransferor rules."
```
Sorry to report this, but can you please properly space these strings concatenation? Otherwise the error message in the logs won't look great.

amaltaro reviewed (commented):
@hassan11196 as discussed in the WM team meeting today, we will proceed with this functional patch, and follow up with enhancements to avoid running the same thing over and over in a future ticket, to be created by @todor-ivanov .

I left only one comment along the code, other than that (and the need to squash all these commits), it all looks good. Please update the PR description (title/commit if needed, too).

Would you be able to also update this documentation: https://gitlab.cern.ch/dmwm/wmcore-docs/-/blob/master/docs/microservices/ReqMgr2-MicroService-RuleCleaner.md ? If not, someone from WM can update it to reflect these changes.

hassan11196 commented:
Hi @amaltaro 
Thank you for the review, I have updated and squashed the commits, let me know if anything else needs to be updated.
I have requested access to the docs repo on mattermost, I will create a PR there as well. 

Thanks.

hassan11196 commented:
PR for documentation update, 
https://gitlab.cern.ch/dmwm/wmcore-docs/-/merge_requests/86

amaltaro reviewed (approved):
Thank you for applying the necessary changes and updating the documentation, @hassan11196 . It all looks good to me.

I will later today backport it to the branch `2.3.8_cmsweb` and make a new patch release.
